vendor/brave-ios/Ledger/Data/BATLedgerDatabase.mm

```diff
@@ -627,4 +651,223 @@ + (double)reservedAmountForPendingContributions
+}
+
++ (void)insertOrUpdateServerPublisherList:(NSArray<BATServerPublisherInfo *> *)list context:(NSManagedObjectContext *)context
+{
+  for (BATServerPublisherInfo *info in list) {
+    auto pi = [[ServerPublisherInfo alloc] initWithEntity:[NSEntityDescription entityForName:NSStringFromClass(ServerPublisherInfo.class) inManagedObjectContext:context] insertIntoManagedObjectContext:context];
+    pi.publisherID = info.publisherKey;
+    pi.status = static_cast<int32_t>(info.status);
+    pi.excluded = info.excluded;
+    pi.address = info.address;
+    [self insertOrUpdateBannerForServerPublisherInfo:pi banner:info.banner context:context];
+  }
+}
+
++ (void)clearAndInsertList:(NSArray<BATServerPublisherInfo *> *)list
+                completion:(nullable BATLedgerDatabaseWriteCompletion)completion
+{
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    const auto batchSize = 1000.0;
+    const auto numberOfSplits = (NSInteger)ceil(list.count / batchSize);
+    NSMutableArray *split = [[NSMutableArray alloc] init];
+    for (NSUInteger i = 0; i < numberOfSplits; i++) {
+      auto index = i * batchSize;
+      NSArray* sub = [[list subarrayWithRange:NSMakeRange(index, MIN(batchSize, list.count - index))] copy];
+      [split addObject:sub];
+    }
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Splitting " << list.count << " records into " << numberOfSplits << " batches" << std::endl;
+    
+    const auto context = [DataController newBackgroundContext];
+    
+    NSDate *start = [NSDate date];
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Deleting publisher list" << std::endl;
+    [context performBlockAndWait:^{
+      const auto fetchRequest = ServerPublisherInfo.fetchRequest;
+      fetchRequest.entity = [NSEntityDescription entityForName:NSStringFromClass(ServerPublisherInfo.class)
+                                        inManagedObjectContext:context];
+      NSError *error = nil;
+      const auto deleteRequest = [[NSBatchDeleteRequest alloc] initWithFetchRequest:fetchRequest];
+      [context executeRequest:deleteRequest error:&error];
+      
+      if (context.hasChanges) {
+        [context save:nil];
+      }
+    }];
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Deleted publisher list in " << [[NSDate date] timeIntervalSinceDate:start] << " seconds" << std::endl;
+    const auto operationQueue = [[NSOperationQueue alloc] init];
+    // 2 concurrent ops operate better than 4 or 6 when testing
+    operationQueue.maxConcurrentOperationCount = MIN(2, [[NSProcessInfo processInfo] activeProcessorCount]);
+    operationQueue.name = @"Publisher List Insertions";
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Inserting " << list.count << " publisher info objects" << std::endl;
+    for (NSArray *section in split) {
+      [operationQueue addOperationWithBlock:^{
+        const auto newContext = [DataController newBackgroundContext];
+        [newContext performBlockAndWait:^{
```

Comment:
Should this really perform and wait if we want it to run concurrently? It will block one of the threads on the queue.

`DispatchQueue.concurrentPerform` or similar might perform the best here as well or just a dispatch group since order shouldn't matter.

GCD is preferable over operation queues imo.

Reply:
We want the operation to be blocking so that it has to complete that 1000 before running another operation, otherwise the maximum concurrent operations wouldn't mean anything, since the block would return and "complete" immediately. 

I also tried using more threads (which `concurrentPerform` would end up doing) and it actually slowed the overall task down and had consequences on overall browser performance. An iPhone X using 2 threads for instance took on avg 30-40s to complete, while the same device using 4 threads took 40-60s on average

Reply:
What I meant was:

instead of creating an operation queue and calling performAndWait which would only block all the threads in the underlying queue, we could just dispatch all of them to a group and call dispatch wait to wait until ALL are finished.. but they won't block the worker threads and would dispatch asynchronously if serial execution isn't needed.

That's weird that it caused performance issues. It's supposed to be more performant and load-balancing among the thread pool.

Reply:
Yeah but we technically want to limit the number of threads it spawns due to performance concerns mentioned above, otherwise we're going to be dispatching 297 large tasks all at once. We don't necessarily care about serial execution because like you said before the order doesn't matter. The `performBlockAndWait` here is only going to block the private queue created for the operation by `NSOperationQueue`, not all the queues

Reply:
`NSOperationQueue` uses GCD under the hood as far as I'm aware